pkg/rng/entropy.go

```diff
@@ -0,0 +1,160 @@
+	RandomDevice = "/dev/random"
+	// HwRandomDevice is the linux hw random device
+	HwRandomDevice = "/dev/hwrng"
+)
+
+// Can be extended but keep in mind to priorize
+// more secure random sources like hw random over
+// timer, jitter based mechanisms. At the top of the array
+// is the highest priority.
+// <rng-name>
+var trngList = []string{
+	"tpm-rng",
+	"intel-rng",
+	"amd-rng",
+	"timeriomem-rng",
+}
+
+// Searches for available True Random Number Generator
+// inside the kernel api and sets the most secure on if
+// available which seeds /dev/hwrng
+func setAvailableTRNG() (bool, error) {
+	var (
+		currentRNG    string
+		availableRNGs []string
+		selectedRNG   string
+	)
+
+	availableFileData, err := ioutil.ReadFile(HwRandomAvailableFile)
+	if err != nil {
+		return false, err
+	}
+	availableRNGs = strings.Split(string(availableFileData), " ")
+
+	for _, trng := range trngList {
+		for _, rng := range availableRNGs {
+			if trng == rng {
+				selectedRNG = trng
+				break
+			}
+		}
+	}
+
+	if selectedRNG == "" {
+		return false, errors.New("No TRNG found on platform")
+	}
+
+	if err = ioutil.WriteFile(HwRandomCurrentFile, []byte(selectedRNG), 0644); err != nil {
+		return false, err
+	}
+
+	currentFileData, err := ioutil.ReadFile(HwRandomCurrentFile)
+	if err != nil {
+		return false, err
+	}
+	currentRNG = string(currentFileData)
+
+	if currentRNG != selectedRNG {
+		return false, errors.New("Couldn't select TRNG: " + currentRNG)
+	}
+
+	return true, nil
+}
+
+// UpdateLinuxRandomness seeds random data from
+// /dev/hwrng into /dev/random based on a timer and
+// the entropy pool size
+// Usage:
+// go UpdateLinuxRandomness()
+func UpdateLinuxRandomness(recoverer recovery.Recoverer) error {
+	if good, err := setAvailableTRNG(); err != nil {
+		return err
+	} else if !good {
+		return errors.New("Could not find a good TRNG")
+	}
+
+	hwRng, err := os.OpenFile(HwRandomDevice, os.O_RDONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	rng, err := os.OpenFile(RandomDevice, os.O_APPEND|os.O_WRONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer hwRng.Close()
+		defer rng.Close()
+
+		for {
+			time.Sleep(EntropyFeedTime * time.Second)
+
+			randomEntropyAvailableData, err := ioutil.ReadFile(RandomEntropyAvailableFile)
+			if err != nil {
+				recoverer.Recover("Can't read entropy pool size")
+			}
+
+			formatted := strings.TrimSuffix(string(randomEntropyAvailableData), "\n")
+			randomEntropyAvailable, err := strconv.ParseUint(formatted, 10, 32)
+			if err != nil {
+				recoverer.Recover("Can't parse entropy pool size")
+			}
+
+			if randomEntropyAvailable >= EntropyThreshold {
+				continue
+			}
+
+			var random = make([]byte, EntropyBlockSize)
+			length, err := hwRng.Read(random)
+			if err != nil {
+				recoverer.Recover("Can't open the hardware random device")
+			}
+			written, err := rng.Write(random[:length])
+			if err != nil || written != length {
```

Comment:
the check `written != length` is not necessary because "Write returns a non-nil error when n != len(b)." as per  https://golang.org/pkg/os/#File.Write . You should also append the error string to the message

Reply:
Done